### ISSUE TITLE
og:image not taken into account / problem prerendered page / user-agent specific? / Add link cart not available in the interface

### ISSUE THREAD

**Issue by eliselavy** [bluesky-social/social-app#4696]
**Describe the bug**
Link cart not rendered in the social post correcly

The link cart should be: https://raw.githubusercontent.com/jdh-observer/jdh002-m734RWDSLo9C/main/socialmediacover.png
But it's the one of the website: https://github.com/C2DH/journal-of-digital-history/blob/develop/public/img/issues/social-media/jdh003.png

See here: https://bsky.app/profile/jdighist.bsky.social/post/3kvvj33vhuz2q
Call the page: https://journalofdigitalhistory.org/en/article/m734RWDSLo9C
Prerendered page: https://journalofdigitalhistory.org/prerendered/en/article/m734RWDSLo9C

Open graph protocol:
<meta property="og:image" content="https://raw.githubusercontent.com/jdh-observer/jdh002-m734RWDSLo9C/main/socialmediacover.png">
<meta property="og:url" content="https://journalofdigitalhistory.org/en/article/m734RWDSLo9C">

**Expected behavior**

Take into account the image defined in the og:image of the prerendered page 

**Screenshots**

Here an overview of the open graph property
<img width="1120" alt="Screenshot 2024-06-28 at 12 11 49" src="https://github.com/bluesky-social/social-app/assets/35602279/1b9a86ef-cac6-481d-a21a-808ef5a39781">

In the interface no possibility to add link cart
<img width="572" alt="Screenshot 2024-06-28 at 12 10 07" src="https://github.com/bluesky-social/social-app/assets/35602279/fd658b51-705b-41b8-b667-b0e0225215d4">

Is it a problem of nginx properly defined? Is the for bluesky a user-agent specific to add?

```
location ~ ^/en/article/(.*)$ {
    add_header              X-Prerendering upstream-backend always;
    # long awaiting, prerendering with django
    set $prerender 0;

    if ($http_user_agent ~* "linkedinbot|slackbot|googlebot|yahoo|bingbot|baiduspider|yandex|yeti|yodaobot|gigabot|ia_archiver|facebookexternalhit|twitterbot|developers\.google\.com") {
      set $prerender 1;
    }

    if ($prerender = 1) {
      rewrite ^/en/(.*)$ /prerendered/en/$1 last;
    }
    root /opt/app;
    try_files $uri $uri/ /index.html = 404;
  }
```

**Comment by danieleguido** [comment]
Indeed, we need to add Bluesky user agent to the list, yes. The problem is that there is no specific documentation on the user agent string used by Bluesky, so it's pending @todo

**Comment by haileyok** [comment]
@danieleguido let me pass this along and see what we want to do, will ping you again whenever we have something.

**Comment by eliselavy** [comment]
@haileyok any news?

**Comment by eliselavy** [comment]
Ok user-agent called bluesky 

https://webhook.site/#!/view/87a86e39-3c25-427b-9c5d-11286092bc85/ef879487-cb50-4074-a8fe-67523a8330d3/1

**Comment by eliselavy** [comment]
> Ok user-agent called bluesky
> 
> https://webhook.site/#!/view/87a86e39-3c25-427b-9c5d-11286092bc85/ef879487-cb50-4074-a8fe-67523a8330d3/1

@haileyok by using API doesn't work , other user-agent?

**Comment by boogheta** [comment]
Hey @danieleguido we encounter the same issue on some of our websites since some time, did you happen to end up finding a proper solution?

**Comment by salaun-marion** [comment]
> Hey [@danieleguido](https://github.com/danieleguido) we encounter the same issue on some of our websites since some time, did you happen to end up finding a proper solution?

What is exactly the problem you are encountering ?

**Comment by boogheta** [comment]
Basically, preview cards on Bluesky do not display images anymore whereas it used to fully work while we did not change anything with our meta fields.
CardyB now returns an error "Unable to serve image from URL" for our card images: https://cardyb.bsky.app/v1/image?url=https%3A%2F%2Fprojet-resin.fr%2Fdata%2Fimages%2Fcapture-ecran-annuaire-resin.png
I described it a bit more fully here: https://github.com/medialab/website/issues/579

**Comment by salaun-marion** [comment]
@boogheta Ok I got it. We develop a module to post on Bluesky recent JDH article. The stuff is for us, it is working because we are pulling the image from JDH to give to Bluesky and not the opposite. So I am not sure, I can help you with this issue 😔

**Edit**
@boogheta Actually I think I get better your issue now after reading it again. Sorry I misunderstood ^^

In case it can help you, we use this piece of code to fetch the image in our back-end (link for the [`bluesky.py` file](https://github.com/C2DH/journal-of-digital-history-backend/blob/85a65e489aca39d28bd0c8ad6d5bd03976485ed0/jdhapi/utils/bluesky.py#L24)). I have rechecked one hour ago, and it is still working for us.

```
BROWSER_UA = (
    "Mozilla/5.0 AppleWebKit/537.36 (KHTML, like Gecko; compatible; Bluesky Cardyb/1.1; "
    "+mailto:support@bsky.app) Chrome/W.X.Y.Z Safari/537.36"
)

...

def fetch_link_metadata(url: str):
    headers = {"User-Agent": BROWSER_UA}
    r = requests.get(url, headers=headers, timeout=10)
    if r.status_code != 200:
        raise ValueError(f"Failed to fetch URL {url}: {r.status_code}")

    soup = BeautifulSoup(r.text, "html.parser")
    title_tag = soup.find("meta", property="og:title") or soup.find("title")
    desc_tag = soup.find("meta", property="og:description") or soup.find(
        "meta", attrs={"name": "description"}
    )
    img_tag = soup.find("meta", property="og:image")
    title = (
        title_tag["content"]
        if title_tag and title_tag.has_attr("content")
        else (title_tag.text.strip() if title_tag else "")
    )
    description = (
        desc_tag["content"] if desc_tag and desc_tag.has_attr("content") else ""
    )
    image_url = img_tag["content"] if img_tag and img_tag.has_attr("content") else None
    return title, description, image_url
```

In our nginx config file we are serving a prerendered version to Bluesky bot which contains this `og:image` corresponding to our `socialmediacover` file.

And for the bluesky user agent we use it in this string : 
```
    if ($http_user_agent ~* "bluesky|linkedinbot|slackbot|googlebot|yahoo|bingbot|baiduspider|yandex|yeti|yodaobot|gigabot|ia_archiver|facebookexternalhit|twitterbot|developers\.google\.com")
```
